### DISCUSSION TITLE
Making the sun and moon card display consistent

### DISCUSSION THREAD

**Discussion by elydpg** [breezy-weather/breezy-weather#1106] [Ideas]
### Describe your idea

This is a more general idea outlining the inconsistencies in the way that the data in the "Sun and moon" card is presented. 

As a normal, working as intended example, here is midday in Tel Aviv, Israel:  
![20240609_124459](https://github.com/breezy-weather/breezy-weather/assets/8622264/e6865033-3582-4b77-a7fd-7cbe57adaa22)

Here is Toronto, Canada in the early morning before sunrise and moonrise. As I would expect, the sun and moon simply do not appear:  
![20240609_114604](https://github.com/breezy-weather/breezy-weather/assets/8622264/1e197c0a-d90d-4457-9a6b-377a31091245)

Here is Auckland, New Zealand after sunset and after moonset. I would have expected it to look as in the previous picture, but instead the sun and moon are stuck on the end of the screen:  
![20240609_115728](https://github.com/breezy-weather/breezy-weather/assets/8622264/7bb4f149-7e82-45a6-8fae-6cdbd20ee3a7)

And here is Adelaide, Australia after sunset but before moonset. Same issue but it illustrates that moon and sun are getting stuck independently:  
![20240609_115749](https://github.com/breezy-weather/breezy-weather/assets/8622264/1dff3842-0117-44fa-8928-45f194a96c4f)

And then there's the whole mess of what happens in the high arctic. There seems to be no consistency here location to location.

In Svalbard, it shows sunrise and sunset at 01:00, and the sun doesn't even render for some reason. Moonrise and moonset appear from midnight to 23:59:  
![20240609_123132](https://github.com/breezy-weather/breezy-weather/assets/8622264/92d94cd5-1eb4-4650-91d8-1721eed66b0e)

In Alert, Canada it shows sunrise and sunset at 19:00, and moonrise and moonset are as in Svalbard (Note that I'm using Open Meteo here and not ECCC, see #1105). Weirdly the sun path renders but no sun once again and it's put in night mode, even though it's the arctic and the sun is shining:  
![20240609_123154](https://github.com/breezy-weather/breezy-weather/assets/8622264/3a389c9b-9636-4f0f-b4f2-98e8c16a892f)

Finally this is I believe Barrow, Alaska, USA. Sun and moon are both from 00:00 to 23:59 and are rendering in the exact same spot:
![20240609_123202](https://github.com/breezy-weather/breezy-weather/assets/8622264/19618e5d-95dd-4938-9373-1d712626e402)

I think this is largely caused by the app assuming that every calendar day has exactly one sunrise and one sunset, in that order, which is a false assumption. It's also false for moonrise and moonset obviously, which will probably be something that more users experience especially as we head into the first quarter.

I'm not sure exactly how to fix this in code, but here are all of the logical possibilities for a given calendar day:

1. There are no sunsets or sunrises. In this case, it needs to be determined another way whether it is day or night in order to apply the theme accordingly. It might not make sense to render the sun in this case, even during daytime. 
2. There is exactly one sunrise. In this case the time before sunrise is night time and no sun should be rendered. After sunrise, render the sun in position based on the time until the earliest sunset in the next calendar day. 
3. There is exactly one sunset. In this case the time after sunset is night time, so no sun render, and before sunset render the sun in position based on the time since the latest sunrise in the previous calendar day.
4. There is sunrise followed by sunset. This is the typical case; after sunset and before sunrise is night time, in between is daytime. Render the sun accordingly. 
5. There is sunset followed by sunrise. In this case the time between sunset and sunrise is night time so no render. The time before sunset is daytime so do as in case 3, and the time after sunrise is also daytime so do as in case 2
6. There is sunset followed by sunrise followed by sunset. In this case do as in case 3 before the first sunset, otherwise do as in case 4. 
7. There is sunrise followed by sunset followed by sunrise. In this case do as in case 2 after the last sunrise, otherwise do as in case 4. 

The same logic applies to the moonrise and moonset, with the exception that there is no need to determine the theme, only the moon render position. I'm fairly sure as well that cases 6 and 7 don't happen for the moon, as the period between moonrises and moonsets is always about 25 hours. But it never hurts to be fully comprehensive 😅

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.4 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
You're mixing too many things, please open an issue or idea for each thing that needs or you suggest to correct/change. Some are sources-side errors that may need to be workarounded, some are specific to full night or full sun, some might be undocumented normal behaviors. In either case, they won't be solved at the same time, so they need to be handled separately. Please.

**Reply by elydpg** [reply]
If not one big issue, then I can think of four separate issues (these are not fully documented they're just tentative titles):

[Issue] Sun and moon not rendering correctly after setting
[Issue] Moon not appearing after midnight if moonset is after midnight
[Issue] Rise and set times are broken in the high arctic (one issue per source)
[Idea] Use a consistent sun and moon logic regardless of source

Would it be better if I created these separately? (Once I properly documented them of course)

**Reply by papjul** [reply]
Yes, much better!

**Comment by papjul** [comment]
Quick notes:

> the app assuming that every calendar day has exactly one sunrise and one sunset, in that order, which is a false assumption. It's also false for moonrise and moonset obviously, which will probably be something that more users experience especially as we head into the first quarter.

I don't understand why it's a false assumption? When the sun rises, it will always set at some point. Except when it's always up, so we default to 00:00 to 23:59 to represent that. And the next day, 00:00 to 23:59 shows again, meaning it has always been up. As far as I know, all sources give us the closest (sun or moon) rise time after midnight, and then the set time after that rise time (not the closest after 00:00 which could end up before the rise time for the moon, which would make no sense to set before rising). Our own computing when sources don't provide this data (or when we didn't implement it yet, such as ECCC) also work that way.


Regarding sun not showing, it's an Open-Meteo bug, see here for Resolute: https://api.open-meteo.com/v1/forecast?timezone=auto&timeformat=unixtime&latitude=74.69736&longitude=-94.83062&models=best_match&daily=sunrise%2Csunset&forecast_days=16&past_days=1&windspeed_unit=ms

As you can see, according to this data, the sun is rising at `0` and setting at `0`. `0` is the value representation of 1970-01-01 00:00 UTC. So actually what you're seeing is 19:00 on 1969-12-31 (in your location timezone), so of course you won't see the sun rise because the circle takes into account the day, not just hours, so it's completely off. It's also setting at the same time it's rising, so it doesn't make any sense.

I have no way to differentiate "always up" and "always down" if Open-Meteo always gives me the `0` value in both cases. Please report an issue on Breezy Weather repo and I will see if there is any open issue on Open-Meteo side, or open one if needed.

**Reply by elydpg** [reply]
> I don't understand why it's a false assumption? When the sun rises, it will always set at some point. Except when it's always up, so we default to 00:00 to 23:59 to represent that. 

There are some cases where sunset is well after midnight in summer due to timezone quirkiness, and sometimes this can cause days with sunrise but no sunset. For example, in [Nome, Alaska, on May 10, 2024](https://www.timeanddate.com/sun/usa/nome?month=5&year=2024) there was a sunrise at 05:56 but no sunset, as that happened at 00:01 the next day. 

> As far as I know, all sources give us the closest (sun or moon) rise time after midnight, and then the set time after that rise time (not the closest after 00:00 which could end up before the rise time for the moon, which would make no sense to set before rising). Our own computing when sources don't provide this data (or when we didn't implement it yet, such as ECCC) also work that way.

What happens when trying to get the sunrise sunset data for [Vostok Station, Antarctica](https://www.timeanddate.com/sun/antarctica/vostok-station?month=2&year=2024) on February 21, 2024? In this case, there is a sunset followed by a sunrise, but the sunrise is actually at 23:43 after the sunset at 22:31, which means the sunrise is actually the sunrise for the 22nd not the 21st. 

> I have no way to differentiate "always up" and "always down" if Open-Meteo always gives me the `0` value in both cases. Please report an issue on Breezy Weather repo and I will see if there is any open issue on Open-Meteo side, or open one if needed.

It turns out that Open-Meteo has an "is-day" parameter that might help disambiguate [Example for Alert, Canada](https://api.open-meteo.com/v1/forecast?latitude=82.5178&longitude=-62.2806&current=is_day) but I'm not sure that other sources have it 🤷

**Reply by papjul** [reply]
> There are some cases where sunset is well after midnight in summer due to timezone quirkiness, and sometimes this can cause days with sunrise but no sunset. For example, in [Nome, Alaska, on May 10, 2024](https://www.timeanddate.com/sun/usa/nome?month=5&year=2024) there was a sunrise at 05:56 but no sunset, as that happened at 00:01 the next day.

Ah? We are not supposed to have no sunset. If you see any example, then it's an issue to be reported. If a sunset ends up at 24:01, then we will show "00:01" (not many people are familiar with [times counted past the 24-hour mark](https://en.wikipedia.org/wiki/24-hour_clock#Times_after_24:00), so we show "00:01"), but it's still showing even if it is the next day.

> What happens when trying to get the sunrise sunset data for [Vostok Station, Antarctica](https://www.timeanddate.com/sun/antarctica/vostok-station?month=2&year=2024) on February 21, 2024? In this case, there is a sunset followed by a sunrise, but the sunrise is actually at 23:43 after the sunset at 22:31, which means the sunrise is actually the sunrise for the 22nd not the 21st.

Yes, that's correct, the sun will rise at 23:43 and set at 46:31. Since we don't use the 24+ hour notation, you can easily guess it's the next day, as the sun cannot rise at 23:43 and set at 22:31 the same day, since that would be before it rose.

So in the end, we will show 23:43 and 22:31, just like on your link, as that's what people are familiar with.

> It turns out that Open-Meteo has an "is-day" parameter that might help disambiguate [Example for Alert, Canada](https://api.open-meteo.com/v1/forecast?latitude=82.5178&longitude=-62.2806&current=is_day) but I'm not sure that other sources have it 🤷

I don't consider this a reliable way to handle this. Other sources don't have this issue, so we can focus on a way to solve it on Open-Meteo-only (or rather, get it fixed upstream, for example having `null` values when never rising/settings, and 00:00 to 23:59 when up all day (the day we requested) although that last one might not be accepted, but as long as I can differentiate with `null` values, that's okay ;)).

**Comment by elydpg** [comment]
> Yes, that's correct, the sun will rise at 23:43 and set at 46:31. Since we don't use the 24+ hour notation, you can easily guess it's the next day, as the sun cannot rise at 23:43 and set at 22:31 the same day, since that would be before it rose.

So you're saying that Sunrise is at 23:43 on Feb 21, 2024, and then sunset for that day is at 46:31? Not sure that checks out, since that sunset is actually at 22:31 on Feb 21, and the sunset following the 23:43 sunrise is actually at 45:58 on Feb 21, which is to say, 21:58 on the 22nd. So does the 22:31 sunset just get skipped then? 

I found another weird Vostok Antarctica edge case. On Oct 18, 2024, there is a sunrise at 00:18 and a sunset at 21:13. There is then another sunrise at 23:59, and then the next sunset on Oct 19 is on 21:36, but then the next sunrise is on Oct 19 at 23:35. Then on Oct 20 there is a sunset at 22:22 and a sunrise at 22:50. So then what is the sunrise for Oct 18? What about Oct 19 and Oct 20?

> I don't consider this a reliable way to handle this. Other sources don't have this issue, so we can focus on a way to solve it on Open-Meteo-only (or rather, get it fixed upstream, for example having `null` values when never rising/settings, and 00:00 to 23:59 when up all day (the day we requested) although that last one might not be accepted, but as long as I can differentiate with `null` values, that's okay ;)).

Yeah same here. Open-Meteo definitely should be having `null` values instead of unix time zero, especially since Open-Meteo offers historical data so zero (1970-01-01T00:00:00Z) is a valid sunrise or sunset time. That's definitely Open-Meteo's problem though

**Comment by papjul** [comment]
> So you're saying that Sunrise is at 23:43 on Feb 21, 2024, and then sunset for that day is at 46:31? Not sure that checks out, since that sunset is actually at 22:31 on Feb 21, and the sunset following the 23:43 sunrise is actually at 45:58 on Feb 21, which is to say, 21:58 on the 22nd. So does the 22:31 sunset just get skipped then?

Alright, my bad, bad understanding. Then, I believe it depends on the source. I believe most of them would give you the following:
- Sunrise: 23:43
- Sunset: 21:58 D+1

> I found another weird Vostok Antarctica edge case. On Oct 18, 2024, there is a sunrise at 00:18 and a sunset at 21:13. There is then another sunrise at 23:59, and then the next sunset on Oct 19 is on 21:36, but then the next sunrise is on Oct 19 at 23:35. Then on Oct 20 there is a sunset at 22:22 and a sunrise at 22:50. So then what is the sunrise for Oct 18? What about Oct 19 and Oct 20?

Once again, depends on the data given by the source.

I believe our own computing would give you the following:
Oct 18: 00:18 / 21:13
Oct 19: 23:35 (since that's the first sunrise we find after midnight) / 22:22 D+1
Oct 20: 22:50 / Sunset the next day

I already described this behavior here:
> As far as I know, all sources give us the closest (sun or moon) rise time after midnight, and then the set time after that rise time

I'm not sure I want to spend hours discussing edge-cases about isolated places in the world, it's just a hobby project that I'm doing for me and my friends. Don't take it personally if I don't reply swiftly ;)

**Comment by elydpg** [comment]
> I'm not sure I want to spend hours discussing edge-cases about isolated places in the world, it's just a hobby project that I'm doing for me and my friends. Don't take it personally if I don't reply swiftly ;)

Fair enough! You reply quicker than most app developers out there :) I'm kinda a nerd about this stuff so the attention to detail here is something I like a lot (which seems to be a theme of this app, which is perhaps why I'm so enamoured by it!). Will spin the sun moon discussion into the separate issues now
